Source/WebGPU/WebGPU/BindGroup.mm

```diff
@@ -77,34 +110,39 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
-
-            } else if (samplerIsPresent) {
-                id<MTLSamplerState> sampler = WebGPU::fromAPI(entry.sampler).samplerState();
-                ASSERT(sizeof(float*) == sizeof(sampler.gpuResourceID));
-                *(MTLResourceID*)(argumentBufferContents[stage] + offset) = sampler.gpuResourceID;
-
-            } else if (textureViewIsPresent) {
-                id<MTLTexture> texture = WebGPU::fromAPI(entry.textureView).texture();
-                ASSERT(sizeof(float*) == sizeof(texture.gpuResourceID));
-                *(MTLResourceID*)(argumentBufferContents[stage] + offset) = texture.gpuResourceID;
-                resources.append({ texture, MTLResourceUsageRead, renderStage(stage) });
+            auto stages = bindGroupLayout ? bindGroupLayout->stagesForBinding(entry.binding) : (WGPUShaderStage_Vertex | WGPUShaderStage_Fragment | WGPUShaderStage_Compute);
+            for (size_t currentStage = 0; currentStage < shaderStageLength; ++currentStage) {
+                WGPUShaderStage renderStage = shaderStage[currentStage];
+                if (!(stages & renderStage))
+                    continue;
+
+                auto& offset = bindingOffsetForStage[currentStage];
+                ASSERT(argumentBufferContents[currentStage]);
+
+                if (bufferIsPresent) {
+                    id<MTLBuffer> buffer = WebGPU::fromAPI(entry.buffer).buffer();
+                    if (entry.offset > buffer.length)
+                        return BindGroup::createInvalid(*this);
+
+                    ASSERT(sizeof(float*) == sizeof(buffer.gpuAddress));
+                    *(float**)(argumentBufferContents[currentStage] + offset) = (float*)((char*)buffer.gpuAddress + entry.offset);
+                    offset += sizeof(MTLResourceID);
+                } else if (samplerIsPresent) {
+                    id<MTLSamplerState> sampler = WebGPU::fromAPI(entry.sampler).samplerState();
+                    ASSERT(sizeof(float*) == sizeof(sampler.gpuResourceID));
+                    *(MTLResourceID*)(argumentBufferContents[currentStage] + offset) = sampler.gpuResourceID;
+                    offset += sizeof(MTLResourceID);
+                } else if (textureViewIsPresent) {
+                    id<MTLTexture> texture = WebGPU::fromAPI(entry.textureView).texture();
+                    ASSERT(sizeof(float*) == sizeof(texture.gpuResourceID));
+                    *(MTLResourceID*)(argumentBufferContents[currentStage] + offset) = texture.gpuResourceID;
+                    resources.append({ texture, MTLResourceUsageRead, renderStage });
+                    offset += sizeof(MTLResourceID);
+                }
             }
         }
 
-        return BindGroup::create(vertexArgumentBuffer, fragmentArgumentBuffer, computeArgumentBuffer, WTFMove(resources), *this);
+        return BindGroup::create(argumentBuffer[0], argumentBuffer[1], argumentBuffer[2], WTFMove(resources), *this);
```

Comment:
I won't be opposed to something like `argumentBuffer[vertexStage]` instead of `argumentBuffer[0]`.

Reply:
I will make this change during the PR for https://bugs.webkit.org/show_bug.cgi?id=250123 which I am working on now